File: myLeetCode/Dynamic_Programming/House_Robber/198. House Robber.cpp
// 198. House Robber
// https://leetcode.com/problems/house-robber/

#include <iostream>
#include <vector>
#include <algorithm>

int rob(std::vector<int>& nums) {
    int n = nums.size();
    if (n == 0) {
        return 0;
    } else if (n == 1) {
        return nums[0];
    }
    
    std::vector<int> dp(n);
    dp[0] = nums[0];
    dp[1] = std::max(nums[0], nums[1]);
    
    for (int i = 2; i < n; i++) {
        dp[i] = std::max(dp[i-2] + nums[i], dp[i-1]);
    }
    
    return dp[n-1];
}

int main() {
    std::vector<std::vector<int>> testCases = {
        {1, 2, 3, 1},
        {2, 7, 9, 3, 1},
        {1, 2, 3, 4, 5, 6}
    };

    for (int i = 0; i < testCases.size(); i++) {
        int result = rob(testCases[i]);

        std::cout << "For the test case " << i+1 << ", the maximum amount of money that can be robbed is " << result << "." << std::endl;
    }

    return 0;
}

File: C++/代码随想录/Dynamic_Programming/House_Robber/198. House Robber.py
def rob(nums):
    if not nums:
        return 0
    if len(nums) <= 2:
        return max(nums)

    dp = [0] * len(nums)
    dp[0] = nums[0]
    dp[1] = max(nums[0], nums[1])

    for i in range(2, len(nums)):
        dp[i] = max(dp[i - 1], dp[i - 2] + nums[i])

    return dp[-1]

# 测试用例
print(rob([1, 2, 3, 1]))  # 输出应该是 4
print(rob([2, 7, 9, 3, 1]))  # 输出应该是 12
